### PR TITLE
Add discovery via akka-dns example + exercise on the new down-all SBR strategy

### PR DESCRIPTION
**Pull request by eloots**
Adds SBR exercise with `down-all` strategy
Adds Akka Bootstrap / Discovery using `akka-dns`

### CODE REVIEW

**Comment by raboof** [comment]
From chat: the interesting example here is `exercise_051_`, and the question is why cluster bootstrap appears to be doing `A` lookups instead of `SRV` lookups.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+akka {
+
+  io.dns.resolver = async-dns
+
+  log-config-on-start = on
+  discovery.method = akka-dns
+  management.cluster.bootstrap.contact-point-discovery {
+    service-name = "_picluster._tcp.lb.com"
+    protocol = "tcp"
+    port = "8558"
```
Comment (raboof) [inline exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: This looks weird. Looking at the docs (https://doc.akka.io/docs/akka/current/discovery/index.html#discovery-method-dns), wouldn't it make more sense to have `service-name="lb.com"`, `protocol="tcp"` and `port="picluster"`?

Comment (raboof) [inline exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: 'port' is not a valid option here (https://github.com/akka/akka-management/blob/master/cluster-bootstrap/src/main/resources/reference.conf#L25)

Comment (raboof) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: `port-name` does exist and defaults to `${akka.management.http.port}` which defaults to `8558`, though.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+akka.coordinated-shutdown.exit-jvm = on
+
+akka {
+
+  io.dns.resolver = async-dns
+
+  log-config-on-start = on
+  discovery.method = akka-dns
+  management.cluster.bootstrap.contact-point-discovery {
+    service-name = "_picluster._tcp.lb.com"
```
Comment (raboof) [inline exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: If I'm reading https://doc.akka.io/docs/akka/current/discovery/index.html#discovery-method-dns right, shouldn't this be `service-name = "picluster.lb.com` to get SRV requests for `_8558._tcp_picluster.lb.com` (or `_management._tcp._picluster.lb.com`)?

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 1 above)
Comment (eloots) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: @raboof Indeed: DNS discovery uses `A` records by default (as I can see using WireShark).
I wanted to use `SRV` records as this is to demonstrate it in action and I serve DNS records using `dnsmasq`. The latter doesn't support having multiple IP addresses for a single record, so I had to resort to `SRV` records.
Thanks for chasing this. I think the docs should be more clear about this.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 2 above)
Comment (eloots) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: Indeed: now that I get the mechanism, it all makes sense.
Will adapt accordingly.
Also, verified that this now works.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 1 above)
Comment (raboof) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: > I think the docs should be more clear about this.

Could you please open a ticket and/or a PR?

File: exercise_012_split_brain_resolver_static_quorum_http_mamagement/src/main/scala/org/neopixel/package.scala [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+    def getitem(pos: Int): Long = ws.ws2811_led_get(channel, pos)
+
+    def setitem(pos: Int, value: Long) = {
+      ledData(pos) = value
+      ws.ws2811_led_set(channel, pos, value)
+    }
+  }
+
+  object Adafruit_NeoPixel {
```
Comment (raboof) [inline exercise_012_split_brain_resolver_static_quorum_http_mamagement/src/main/scala/org/neopixel/package.scala]: This seems to be duplicated a lot, perhaps move it to a library? (perhaps in a separate PR)

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  //log-config-on-start = on
+
+  // We'll use Akka Cluster Bootstrap with Discovery via akka-dns
+  discovery.method = akka-dns
+
+  management.cluster.bootstrap.contact-point-discovery {
+    required-contact-point-nr = 3
+    // If you want Akka Discovery via akka-dns to use SRV records,
+    // adapt the following settings. Akka Discovery will look for
+    // SRV entries with path: _management._tcp.picluster.lb.com
```
Comment (raboof) [inline exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]:
```suggestion
    // SRV entries with path: _management._tcp._picluster.lb.com
```
right?

File: exercise_012_split_brain_resolver_static_quorum_http_mamagement/src/main/scala/org/neopixel/package.scala (hunk 3 above)
Comment (eloots) [reply exercise_012_split_brain_resolver_static_quorum_http_mamagement/src/main/scala/org/neopixel/package.scala]: Good point - To be marked in an issue.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 4 above)
Comment (eloots) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: Don't think so (as verified with this config on the queries with WireShark).
The path is `_<port-name>._<protocol>.<service-name>`, so no `_` in fornt of the `service-name`.

**Comment by eloots** [comment]
> there seem to be a lot of different things in this PR apart from just the discovery-dns stuff (renames, changes in different exercises). would have been good to split that up more, now it is hard to follow why changes were made.
> 
> If I understand correctly the discovery-dns example needs a dnsmasq to run, should its configuration also be included?

Doing that - the README.md file for this exercise was missing.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 4 above)
Comment (raboof) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: Interesting! If you are sure, could you update https://github.com/akka/akka/blob/master/akka-docs/src/main/paradox/discovery/index.md#discovery-method-dns accordingly?

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 1 above)
Comment (eloots) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: @raboof I'll open a ticket and see if I can create a PR.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 4 above)
Comment (eloots) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: Will do!

**Comment by eloots** [comment]
> there seem to be a lot of different things in this PR apart from just the discovery-dns stuff (renames, changes in different exercises). would have been good to split that up more, now it is hard to follow why changes were made.

@raboof Correct... I could have separated the addition of the exercise on the `down-all` strategy in a specific PR. As this new exercise is inserted between two existing exercises, the exercises after the new one are renumbered leading to an _explosion_ in changed files because the folder names change. Little that can be done about this unfortunately.
In addition to that, I had already started working on exercise 51 (akka-dns), so this got "bundled" with it.

File: exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf (hunk 4 above)
Comment (eloots) [reply exercise_051_cluster_singleton_akka_bootstrap_dns/src/main/resources/application.conf]: WireShark
